builtin/rm.c

```diff
@@ -262,6 +262,10 @@ int cmd_rm(int argc, const char **argv, const char *prefix)
 	char *seen;
 
 	git_config(git_default_config, NULL);
+	if (the_repository->gitdir) {
+		prepare_repo_settings(the_repository);
+		the_repository->settings.command_requires_full_index = 0;
+	}
```

Comment:
I'm slightly suspicious here.

> You can avoid the need for if (the_repository->gitdir) - which is when the index is needed before parse_options() is called ([link](https://github.com/ffyuanda/git/commit/059fda190215d18e7aa23f825cd607b16a016b65))

In this commit they actually use `if (the_repository->gitdir)` as the way to go:

> However, it is unfortunately not that simple. In `cmd_pack_objects()`,
> for example, the repo settings need to be fully populated so that the
> command-line options `--sparse`/`--no-sparse` can override them, not the
> other way round.
>
> Therefore, we choose to imitate the strategy taken in `cmd_diff()`,
> where we simply do not bother to prepare and initialize the repo
> settings unless we have a `gitdir`.

Though I'm following your advice right now because it does make sense to me ;-)